rqt_py_common/src/rqt_py_common/message_helpers.py

```diff
@@ -0,0 +1,200 @@
+# Copyright (c) 2018, PickNik Robotics
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#
+#   * Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+#   * Redistributions in binary form must reproduce the above
+#     copyright notice, this list of conditions and the following
+#     disclaimer in the documentation and/or other materials provided
+#     with the distribution.
+#   * Neither the name of the PickNik Robotics nor the names of its
+#     contributors may be used to endorse or promote products derived
+#     from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+# Author: Michael Lautman
+
+from ament_index_python import get_resource
+from ament_index_python import get_resources
+from ament_index_python import has_resource
+
+from rclpy import logging
+
+MSG_MODE = 'msg'
+SRV_MODE = 'srv'
+ACTION_MODE = 'action'
+
+# Taken from https://github.com/ros2/ros2cli/blob/bouncy/ros2msg/ros2msg/api/__init__.py
+def get_all_service_types():
+    all_service_types = {}
+    for package_name in get_resources('rosidl_interfaces').keys():
+        service_types = get_service_types(package_name)
+        if service_types:
+            all_service_types[package_name] = service_types
+    return all_service_types
+
+
+def get_service_types(package_name):
```

Comment:
Regarding the above comment, maybe add a 'resources' parameter with a default None value

Reply:
Seems like premature optimization

Reply:
Considering get_resources scales  linearly with the number of packages installed and you are following it up with calls to has_resource and get_resource, it seems like optimization should be considered. These functions require OS calls.

Reply:
Linear isn't a big deal in this case. Even if we have a thousand packages, RQT only needs to run at human speed. I removed the has_resource call and the try except block since I was already catching exceptions thrown by get_resource.